turbopack/crates/turbo-tasks-backend/src/backend/storage.rs

```diff
@@ -119,246 +119,155 @@ impl PersistanceState {
     }
 
     pub fn persistance_state(&self) -> &PersistanceState {
-        match self {
-            InnerStorage::Plain {
-                persistance_state, ..
-            } => persistance_state,
-            InnerStorage::Indexed {
-                persistance_state, ..
-            } => persistance_state,
-        }
+        &self.persistance_state
     }
 
     pub fn persistance_state_mut(&mut self) -> &mut PersistanceState {
-        match self {
-            InnerStorage::Plain {
-                persistance_state, ..
-            } => persistance_state,
-            InnerStorage::Indexed {
-                persistance_state, ..
-            } => persistance_state,
-        }
+        &mut self.persistance_state
     }
 
-    fn check_threshold(&mut self) {
-        let InnerStorage::Plain {
-            map: plain_map,
-            persistance_state,
-        } = self
-        else {
-            return;
-        };
-        if plain_map.len() >= INDEX_THRESHOLD {
-            let mut map: IndexedMap<T> = AutoMap::new();
-            for (key, value) in take(plain_map).into_iter() {
-                let index = key.index();
-                map.entry(index).or_default().insert(key, value);
-            }
-            *self = InnerStorage::Indexed {
-                map,
-                persistance_state: take(persistance_state),
-            };
+    fn get_or_create_map_mut(&mut self, ty: CachedDataItemType) -> &mut CachedDataItemStorage {
+        let i = self.map.iter().position(|m| m.ty() == ty);
+        if let Some(i) = i {
+            &mut self.map[i]
+        } else {
+            self.map.push(CachedDataItemStorage::new(ty));
+            self.map.last_mut().unwrap()
         }
     }
 
-    fn get_or_create_map_mut(&mut self, key: &T::Key) -> &mut AutoMap<T::Key, T::Value> {
-        self.check_threshold();
-        match self {
-            InnerStorage::Plain { map, .. } => map,
-            InnerStorage::Indexed { map, .. } => map.entry(key.index()).or_default(),
-        }
+    fn get_map_mut(&mut self, ty: CachedDataItemType) -> Option<&mut CachedDataItemStorage> {
+        self.map.iter_mut().find(|m| m.ty() == ty)
     }
 
-    fn get_map_mut(&mut self, key: &T::Key) -> Option<&mut AutoMap<T::Key, T::Value>> {
-        self.check_threshold();
-        match self {
-            InnerStorage::Plain { map, .. } => Some(map),
-            InnerStorage::Indexed { map, .. } => map.get_mut(&key.index()),
-        }
+    fn get_map_index(&mut self, ty: CachedDataItemType) -> Option<usize> {
+        self.map.iter_mut().position(|m| m.ty() == ty)
     }
 
-    fn get_map(&self, key: &T::Key) -> Option<&AutoMap<T::Key, T::Value>> {
-        match self {
-            InnerStorage::Plain { map, .. } => Some(map),
-            InnerStorage::Indexed { map, .. } => map.get(&key.index()),
-        }
+    fn get_map(&self, ty: CachedDataItemType) -> Option<&CachedDataItemStorage> {
+        self.map.iter().find(|m| m.ty() == ty)
```

Comment:
Technically this is still constant time, but it might be beneficial to not have always iterate through the list here?

Reply:
We only have a maximum of 30 different types.

According to https://github.com/yegor256/micromap#benchmark using a Vec is faster. That's also the idea behind the `AutoMap` we have

Reply:
@sokra As a micro-optimization, if we create the maps for the most common `CachedDataItemType`s in `InnerStorage::new`, we can ensure they always end up at the start of the list.

However, this may naturally tend to happen anyways (the most frequently accessed types are likely to get initialized first).

Reply:
As follow-up PR I want to pull "common" types out of this vec and just give them a native field in InnerStorage. That would save about 240 bytes per task which is about 11% of total memory